### ISSUE TITLE
TTL parameter required when using multicast across multiple data centres

### ISSUE THREAD

**Issue by chrisjstevo** [real-logic/Aeron#209]
The default ttl for a Udp socket is either 1 or 0, this means that if you have to pass through more than just your local switch, the packets will get rejected. 

Adding this parameter to the channel connection string would allow users to send multicast spanning more than one switch.

> Per-socket/channel seems to make more sense, the use case is cross data centre multicast. We have > a remote backup where the multicast is being routed to. We have several hops to get there. I'm seeing > with aeron I get dropped when I get beyond the first switch, testing with jgroups multicast wrapper (yes > I said the dirty word  ) I can set a ttl of > 1 and the test flows through fine.

> TTL would best be a channel attribute, IMO
> the reasoning behind why it isn't there now is simply haven't gotten to it yet
> probably should add TTL to the canonical form so that different TTLs can be separate channels
> I will add putting this in to my list. @chrisjstevo would you be so kind as to add an issue to capture the request?

**Comment by chrisjstevo** [comment]
For reference, adding the below hack into UdpChannelTransport means my multicast gets routed correctly cross datacentre. 

So adding a TTL definitely allows it to traverse switches.

                receiveDatagramChannel.setOption(StandardSocketOptions.SO_REUSEADDR, true);
                receiveDatagramChannel.bind(new InetSocketAddress(endPointAddress.getPort()));
                receiveDatagramChannel.join(endPointAddress.getAddress(), localInterface);
                sendDatagramChannel.setOption(StandardSocketOptions.IP_MULTICAST_IF, localInterface);

                //hack
                receiveDatagramChannel.setOption(StandardSocketOptions.IP_MULTICAST_TTL, 25);
                sendDatagramChannel.setOption(StandardSocketOptions.IP_MULTICAST_TTL, 25);

                if (null != connectAddress)
                {
                    sendDatagramChannel.connect(connectAddress);
                }

**Comment by mjpt777** [comment]
We should do this with a global config option that can have channel specific overrides.

**Comment by chrisjstevo** [comment]
In the vain of:

public static final String PUBLICATION_UNBLOCK_TIMEOUT_PROP_NAME = "aeron.publication.unblock.timeout"

so -D param? If so am happy to try a patch out for you to review...

for the override by channel, would you see it being grabbed from the Uri somewhere like this:

//UpdChannel.java:76
            final AeronUri uri = parseIntoAeronUri(uriStr);

            validateConfiguration(uri);

I had a go at adding it here as an extra param earlier, but favoured to filthy hack to make sure it worked:

            return AeronUri.builder()
                .media(UDP_MEDIA_ID)
                .param(GROUP_KEY, group)
                .param(INTERFACE_KEY, inf)
                .newInstance();
        }

?

Chris

**Comment by mjpt777** [comment]
Hi @chrisjstevo Can you move this to another issue so we can track and discuss it separately?

**Comment by mjpt777** [comment]
Sorry @chrisjstevo I misread. I thought you wanted to change PUBLICATION_UNBLOCK_TIMEOUT_PROP_NAME. Duh!

**Comment by chrisjstevo** [comment]
Ha, will kill the new issue.
